fedimint-api/src/db/mod.rs

```diff
@@ -880,3 +942,135 @@ mod tests {
         assert_eq!(returned_keys, expected_keys);
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use std::time::Duration;
+
+    use futures::Future;
+
+    use super::mem_impl::MemDatabase;
+    use super::*;
+
+    async fn future_returns_shortly<F: Future>(sub: F) -> Option<F::Output> {
+        tokio::time::timeout(Duration::from_millis(10), sub)
+            .await
+            .ok()
+    }
+
+    fn waiter(db: &Database, key: TestKey) -> tokio::task::JoinHandle<TestVal> {
+        let db = db.clone();
+        tokio::spawn(async move {
+            db.wait_key(&key, ModuleDecoderRegistry::default())
+                .await
+                .expect("DB error")
+        })
+    }
+
+    #[tokio::test]
+    async fn test_wait_key_before_transaction() {
+        let key = TestKey(1);
+        let val = TestVal(2);
+        let db = Database::from(MemDatabase::new());
+
+        let key_task = waiter(&db, TestKey(1));
+
+        let mut tx = db.begin_transaction(ModuleDecoderRegistry::default()).await;
+        tx.insert_new_entry(&key, &val).await.expect("DB error");
+        tx.notify_key(&key).await;
```

Comment:
As this is RN, I think it has a race condition. If `tx.notify_key(&key).await;` finishes before waiter gets to `db.wait_key`, the test will hang.

Reply:
I think it is fixed now.